server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -95,51 +105,71 @@ private static long[] getIndexBounds(final SearchContext context, final String f
             }
         }
 
-        if (min == Long.MAX_VALUE || max == Long.MIN_VALUE) return null;
-
+        if (min == Long.MAX_VALUE || max == Long.MIN_VALUE) {
+            return null;
+        }
         return new long[] { min, max };
     }
 
     /**
-     * This method also acts as a pre-condition check for the optimization,
-     * returns null if the optimization cannot be applied
+     * This method also acts as a pre-condition check for the optimization
+     *
+     * @return null if the processed query not as expected
      */
-    public static long[] getAggregationBounds(final SearchContext context, final String fieldName) throws IOException {
+    public static long[] getDateHistoAggBounds(final SearchContext context, final String fieldName) throws IOException {
         final Query cq = unwrapIntoConcreteQuery(context.query());
-        final long[] indexBounds = getIndexBounds(context, fieldName);
         if (cq instanceof PointRangeQuery) {
             final PointRangeQuery prq = (PointRangeQuery) cq;
-            // Ensure that the query and aggregation are on the same field
-            if (prq.getField().equals(fieldName)) {
-                return new long[] {
-                    // Minimum bound for aggregation is the max between query and global
-                    Math.max(NumericUtils.sortableBytesToLong(prq.getLowerPoint(), 0), indexBounds[0]),
-                    // Maximum bound for aggregation is the min between query and global
-                    Math.min(NumericUtils.sortableBytesToLong(prq.getUpperPoint(), 0), indexBounds[1]) };
-            }
+            final long[] indexBounds = getIndexBounds(context, fieldName);
+            if (indexBounds == null) return null;
+            return getBoundsWithRangeQuery(prq, fieldName, indexBounds);
         } else if (cq instanceof MatchAllDocsQuery) {
-            return indexBounds;
+            return getIndexBounds(context, fieldName);
+        } else if (cq instanceof FieldExistsQuery) {
+            // when a range query covers all values of a shard, it will be rewrite field exists query
+            if (((FieldExistsQuery) cq).getField().equals(fieldName)) {
+                return getIndexBounds(context, fieldName);
+            }
         }
-        // Check if the top-level query (which may be a PRQ on another field) is functionally match-all
-        Weight weight = context.searcher().createWeight(context.query(), ScoreMode.COMPLETE_NO_SCORES, 1f);
-        for (LeafReaderContext ctx : context.searcher().getIndexReader().leaves()) {
-            if (weight.count(ctx) != ctx.reader().numDocs()) {
+
+        return null;
+    }
+
+    private static long[] getDateHistoAggBoundsSegLevel(final SearchContext context, final String fieldName) throws IOException {
+        final long[] indexBounds = getIndexBounds(context, fieldName);
```

Comment:
Why are we loading the index-wide bounds when we're looking at a single segment? Could we just get the min and max values for the current segment?

Reply:
Here's my old thinking:
Date histogram requires to compute bucket ranges, dynamically by rounding the values.
ex: 2023-2-6 8:23 falls into 2023-2-6 8:00 bucket by hour interval
So I try to always give the same input (shard level bounds) so that to output the same buckets.

However, I just realise segments can be treated separately when we know it's functionally match all. 
And I guess now I understand more about the meaning of "every segment is a lucene index" 🙂